static/shared/js/typeahead.js

```diff
@@ -44,27 +44,10 @@ function query_matches_string(query, source_str, split_char) {
     }
 
     // If there is a separator character in the query, then we
-    // require a perfect prefix match (e.g. for 'ab cd ef',
-    // query needs to be e.g. 'ab c', not 'cd ef' or 'b cd
-    // ef', etc.).
-
-    const queries = query.split(split_char);
-    const sources = source_str.split(split_char);
-    let i;
-
-    for (i = 0; i < queries.length - 1; i += 1) {
-        if (sources[i] !== queries[i]) {
-            return false;
-        }
-    }
-
-    // This block is effectively a final iteration of the last
-    // loop.  What differs is that for the last word, a
-    // partial match at the beginning of the word is OK.
-    if (sources[i] === undefined) {
-        return false;
-    }
-    return sources[i].startsWith(queries[i]);
+    // require the match to start at the start of a token.
+    // (E.g. for 'ab cd ef', query could be 'ab c' or 'cd ef',
+    // but not 'b cd ef'.)
+    return source_str.startsWith(query) || source_str.includes(split_char + query);
```

Comment:
The main function change is this commit. I dug up the history as these commits:
* 542f4766d3a01322a31570c39b73c54ed25dbeeb
* 5edbcb87fd40ef87fc4f72bf96a2c5af4b64f0ba

Reply:
Thanks. It looks like before those two, it was already trying to do something a lot like what it currently does (though the old code there is messy so it's hard to tell); it doesn't look like those contain any history of discussing why this behavior is desirable.

This seems to be the commit that introduced this sort of behavior: 245d65eb9e734dc673956c27105623ae7547a08b. #8279 and #8286 have the discussion.

Reading those, I don't see anything in that discussion that suggests that "Peyton Jones" or "Peyton J" shouldn't match "Simon Peyton Jones". So, AFAICT, that behavior was purely an undesired side effect.

I'll add some history to the commit message.

Reply:
Yeah, that seems right to me as well.  I discussed the change with @alya and neither of us could see a reason it'd be a regression, so I think we can proceed with this.